ext/standard/string.c

```diff
@@ -1030,9 +1030,14 @@ PHP_FUNCTION(implode)
 	ZEND_PARSE_PARAMETERS_START(1, 2)
 		Z_PARAM_ARRAY_HT_OR_STR(arg1_array, arg1_str)
 		Z_PARAM_OPTIONAL
-		Z_PARAM_ARRAY_HT_OR_NULL(pieces)
+		Z_PARAM_ARRAY_HT(pieces)
 	ZEND_PARSE_PARAMETERS_END();
 
+	if (arg1_str != NULL && pieces == NULL) {
+		zend_type_error("%s(): Argument #2 ($array) must be of type array, null given", get_active_function_name());
+		RETURN_THROWS();
+	}
```

Comment:
```suggestion
	if (arg1_str != NULL && pieces == NULL) {
		zend_type_error("%s(): Argument #2 ($array) must be of type array, "
			"if argument #1 ($separator) is given", get_active_function_name());
		RETURN_THROWS();
	}
```

Mentioning ``null`` here doesn't make any sense any more if you remove the possibility to provide ``null`` to ``pieces``.

However, I think an improvement for the error message without changing the current ZPP check would be:

```suggestion
	if (arg1_str != NULL && pieces == NULL) {
		zend_type_error("%s(): Argument #2 ($array) must be of type array, "
			"if argument #1 ($separator) is of type string, null given",
			get_active_function_name());
		RETURN_THROWS();
	}
```
Or maybe flip the order:
```suggestion
	if (arg1_str != NULL && pieces == NULL) {
		zend_type_error("%s(): If argument #1 ($separator) is of type string, "
			"argument #2 ($array) must be of type array, null given",
			get_active_function_name());
		RETURN_THROWS();
	}
```

Reply:
I implemented the last suggestion.
But you said it doesn't make sense to mention `null` but you kept it.

If you don't pass the parameter, it will come as null, isn't it? Otherwise, should the message say the parameter is mandatory os something like that?